tools/distpartitioning/gloo_wrapper.py

```diff
@@ -59,71 +59,84 @@ def alltoall_cpu(rank, world_size, output_tensor_list, input_tensor_list):
-       other processes as part of this exchange
-    """
-    rcv_list = []
-    output_list = [None] * world_size
-    for i in range(world_size):
-        rcv_list.clear()
-        rcv_list.append(None)
-        if (i == rank):
-            dist.scatter_object_list(rcv_list, input_list, src = rank)
-        else:
-            send_list = [None] * world_size
-            dist.scatter_object_list(rcv_list, send_list, src = i)
-        output_list[i] = rcv_list[0]
-
-    return output_list
-
-def alltoallv_cpu(rank, world_size, output_tensor_list, input_tensor_list):
-    """
-    Each process scatters list of input tensors to all processes in a cluster
-    and return gathered list of tensors in output list.
+    Returns:
+    --------
+    list : 
+        list of tensors received from other processes during alltoall message
 
-    Parameters
-    ----------
-    rank : int
-        The rank of current worker
-    world_size : int
-        The size of the entire
-    output_tensor_list : List of tensor
-        The received tensors
-    input_tensor_list : List of tensor
-        The tensors to exchange
     """
-    # send tensor to each target trainer using torch.distributed.isend
-    # isend is async
-    senders = []
-    for i in range(world_size):
-        if i == rank:
-            output_tensor_list[i] = input_tensor_list[i].to(torch.device('cpu'))
+    #ensure len of input_tensor_list is same as the world_size.
+    assert input_tensor_list != None
+    assert len(input_tensor_list) == world_size
+
+    #ensure that all the tensors in the input_tensor_list are of same size.
+    sizes = [list(x.size()) for x in input_tensor_list]
+    for idx in range(1,len(sizes)):
+        assert len(sizes[idx-1]) == len(sizes[idx]) #no. of dimensions should be same
+        assert input_tensor_list[idx-1].dtype == input_tensor_list[idx].dtype # dtype should be same 
+        assert sizes[idx-1][1:] == sizes[idx][1:] #except first dimension remaining dimensions should all be the same
+
+    #decide how much to pad. 
+    #always use the first-dimension for padding. 
+    ll = [ x[0] for x in sizes ]
+
+    #dims of the padding needed, if any
+    #these dims are used for padding purposes.
+    diff_dims = [ [np.amax(ll) - l[0]] + l[1:] for l in sizes ]
+
+    #pad the actual message
+    input_tensor_list = [torch.cat((x, torch.zeros(diff_dims[idx]).type(x.dtype))) for idx, x in enumerate(input_tensor_list)]
+
+    #send useful message sizes to all
+    send_counts = []
+    recv_counts = []
+    for idx in range(world_size):
+        #send a vector, of atleast 3 elements, [a, b, ....] where 
+        #a = useful message dim, b = actual message outgoing message size along the first dimension
+        #and remaining elements are the remaining dimensions of the tensor
+        send_counts.append(torch.from_numpy(np.array([sizes[idx][0]] + [np.amax(ll)] + sizes[idx][1:] )).type(torch.int64))
```

Comment:
i don't think `sizes[idx][1:]` don't need to be sent over. all processes know `sizes[idx][1:]`

Reply:
This is on the sending side, where sizes[idx][:1] is same for all the tensors. But on the receiving side, it receives different size'd tensors from everyone else. So potentially the receiving side will receive tensors of dimensions which WILL NOT be of the same size. And for this reason, we need sizes[idx][:1] to be sent to over from the sending process.

Reply:
Added a comment, in the docstring, to indicate the assumption about this comment.